seccomp/umount.go

```diff
@@ -113,19 +120,237 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
 	return u.tracer.createContinueResponse(u.reqId), nil
 }
 
+// umountAllowed purpose is to prevent immutable resources from being
+// unmounted.
+//
+// Method will return 'true' when the unmount operation is deemed legit, and
+// will return 'false' otherwise.
+func (u *umountSyscallInfo) umountAllowed(
+	mip domain.MountInfoParserIface) (bool, *sysResponse) {
+
+	// Skip file-systems explicitly handled by sysbox-fs.
+	if u.FsType == "proc" || u.FsType == "sysfs" {
+		return true, nil
+	}
+
+	// There must be mountinfo state present for this target. Otherwise, let
+	// kernel handle the error.
+	info := mip.GetInfo(u.Target)
+	if info == nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	//
+	// The following scenarios are relevant within the context of this function
+	// and will be handled separately to ease the logic comprehension and its
+	// maintenability / debuggability.
+	//
+	// The different columns in this table denote the 'context' in which the
+	// unmount process is executing, and thereby, dictates the logic chosen
+	// to handle each unmount request.
+	//
+	//    +-----------+--------------+--------------+----------+
+	//    | Scenarios | Unshare(mnt) | Pivot-root() | Chroot() |
+	//    +-----------+--------------+--------------+----------+
+	//    | 1         | no           | no           | no       |
+	//    | 2         | no           | yes          | no       |
+	//    | 3         | no           | no           | yes      |
+	//    | 4         | no           | yes          | yes      |
+	//    | 5         | yes          | no           | no       |
+	//    | 6         | yes          | yes          | no       |
+	//    | 7         | yes          | no           | yes      |
+	//    | 8         | yes          | yes          | yes      |
+	//    +-----------+--------------+--------------+----------+
+	//
+
+	// Identify the mount-ns of the process launching the unmount to compare it
+	// with the one of the sys container's initpid. In the unlikely case of an
+	// error, let the kernel deal with it.
+	processMountNs, err := u.processInfo.MountNsInode()
+	if err != nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+	initProcMountNs, err := u.cntr.InitProc().MountNsInode()
+	if err != nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	// Obtain the sys-container's root-path inode.
+	syscntrRootInode := u.cntr.InitProc().RootInode()
+
+	// If process' mount-ns matches the sys-container's one, then we can simply
+	// rely on the target's mountID to discern an immutable target from a
+	// regular one. Otherwise, we cannot rely on the mountID field, as the values
+	// allocated by kernel for these very mountpoints will differ in other mount
+	// namespaces.
+	if processMountNs == initProcMountNs {
+
+		if u.processInfo.Root() == "/" {
+			processRootInode := u.processInfo.RootInode()
+
+			// Scenario 1): no-unshare(mnt) & no-pivot() & no-chroot()
```

Comment:
Scenarios 1->4 below result result in the same action: checking if the mount is immutable by ID. Thus, I suggest not differentiating between them and instead merging them into a single call to "IsImmutableMountID". In fact, the call to mip.ExtractInode("/") in scenarios 3 and 4 is expensive (due to the nsenter) and not really necessary.

This will simplify the code: if the process doing the unmount is in the same mount ns as the sys container init process, we simply lookup the mount by ID and we are done.